rapidjson.cpp

```diff
@@ -1890,46 +1939,22 @@ static PyMemberDef decoder_members[] = {
-    0,                                        /* tp_print */
-    0,                                        /* tp_getattr */
-    0,                                        /* tp_setattr */
-    0,                                        /* tp_compare */
-    0,                                        /* tp_repr */
-    0,                                        /* tp_as_number */
-    0,                                        /* tp_as_sequence */
-    0,                                        /* tp_as_mapping */
-    0,                                        /* tp_hash */
-    (ternaryfunc) decoder_call,               /* tp_call */
-    0,                                        /* tp_str */
-    0,                                        /* tp_getattro */
-    0,                                        /* tp_setattro */
-    0,                                        /* tp_as_buffer */
-    Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE, /* tp_flags */
-    decoder_doc,                              /* tp_doc */
-    0,                                        /* tp_traverse */
-    0,                                        /* tp_clear */
-    0,                                        /* tp_richcompare */
-    0,                                        /* tp_weaklistoffset */
-    0,                                        /* tp_iter */
-    0,                                        /* tp_iternext */
-    0,                                        /* tp_methods */
-    decoder_members,                          /* tp_members */
-    0,                                        /* tp_getset */
-    0,                                        /* tp_base */
-    0,                                        /* tp_dict */
-    0,                                        /* tp_descr_get */
-    0,                                        /* tp_descr_set */
-    0,                                        /* tp_dictoffset */
-    0,                                        /* tp_init */
-    0,                                        /* tp_alloc */
-    decoder_new,                              /* tp_new */
-    PyObject_Del,                             /* tp_free */
+static PyType_Slot Decoder_Type_Slot[] = {
+    {Py_tp_doc, const_cast<char*>(decoder_doc)},
+    {Py_tp_call, reinterpret_cast<void*>(decoder_call)},
+    {Py_tp_members, decoder_members},
+    {Py_tp_new, reinterpret_cast<void*>(decoder_new)},
+    {Py_tp_traverse, reinterpret_cast<void*>(decoder_traverse)},
+    {0, NULL}
+};
+
+
+static PyType_Spec Decoder_Type_Spec = {
+    "rapidjson.Decoder",                                                                      /* name */
+    sizeof(DecoderObject),                                                                    /* basicsize */
+    0,                                                                                        /* itemsize */
+    Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE | Py_TPFLAGS_HAVE_GC | Py_TPFLAGS_IMMUTABLETYPE, /* flags */
```

Comment:
The type is flagged as `Py_TPFLAGS_IMMUTABLETYPE` to make this change more "noop-like". All static types are immutable. Decoder type remains immutable after it's transformation into heap-type. This flag can be dropped if it's ok to make Decoder to be closer to regular types defined in python.